File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int n = nums.length , l= 0 ,r =  0 , maxLen = 0, zeroes = 0;
        while(r < n){
            if(nums[r] == 0) zeroes++;
            if(zeroes > k){
                if(nums[l] == 0)zeroes--;
                l++;
            }
            if(zeroes <= k){
                maxLen = Math.max(maxLen, r-l+1);
            }
            r++;
        }
        return maxLen;
    }
}

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
         int n= nums.size() , l= 0 ,r =  0 , maxLen = 0, zeroes = 0;
        
        while(r < n){
            if(nums[r] == 0) zeroes++;
            
            while(zeroes > k){
                if(nums[l] == 0) zeroes--;
                l++;
            }
            if(zeroes <= k) maxLen = max(maxLen , r-l+1);
            r++;
        } 
        return maxLen;
    }
};
/* brute 
int maxLen = 0;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            int zeroes = 0;
            for (int j = i; j < n; j++) {
                if (nums[j] == 0) zeroes++;

                if (zeroes > k) break;

                maxLen = max(maxLen, j - i + 1);
            }
        }

        return maxLen;
        
    // Better 
    int n= nums.size() , l= 0 ,r =  0 , maxLen = 0, zeroes = 0;
        
        while(r < n){
            if(nums[r] == 0) zeroes++;
            while(zeroes > k){
                if(nums[l] == 0) zeroes--;
                l++;
            }
            if(zeroes <=k){
                maxLen = max(maxLen,r-l+1);
                r++;
            }
        }
        return maxLen;
        
        
         */